ably.ts.d

```diff
@@ -0,0 +1,457 @@
+    echoMessages?: boolean;
+    environment?: String;
+    logHandler?: any;
+    logLevel?: any;
+    port?: number;
+    queueMessages?: boolean;
+    restHost?: String;
+    realtimeHost?: String;
+    fallbackHosts?: Array<String>
+    recover?: String
+    tls?: boolean;
+    tlsPort?: number;
+    useBinaryProtocol?: boolean;
+    disconnectedRetryTimeout?: number;
+    suspendedRetryTimeout?: number;
+    channelRetryTimeout?: number;
+    httpOpenTimeout?: number;
+    httpRequestTimeout?: number;
+    httpMaxRetryCount?: number;
+    httpMaxRetryDuration?: number;
+  }
+
+  interface AuthOptions {
+    authCallback?: (data: String | TokenDetails | TokenRequest) => void;
+    authHeaders?: Array<any>;
+    authMethod?: String;
+    authParams?: Array<any>;
+    authUrl?: String;
+    key?: String;
+    queryTime?: boolean;
+    token?: String | TokenDetails;
+    tokenDetails?: TokenDetails;
+    useTokenAuth?: boolean;
+  }
+
+  interface TokenParams {
+    capability: String;
+    clientId?: String;
+    nonce?: String;
+    timestamp: Number;   // Time?
+    ttl: Number;
+  }
+
+  interface CipherParams {
+    algorithm: String;
+    key: any;
+    keyLength: number;
+    mode: String;
+  }
+
+  interface ErrorInfo {
+    code: number;
+    message: String;
+    statusCode: number;
+  }
+
+  interface ConnectionDetails {
+    clientId: String;
+    connectionKey: String;
+    connectionStateTtl: number;
+    maxFrameSize: number;
+    maxInboundRate: number;
+    maxMessageSize: number;
+    serverId: String;
+    maxIdleInterval: number;
+  }
+
+  interface StatsMessageTypes {
+    all: number,
+    messages: number,
+    presence: number;
+  }
+
+  interface StatsRequestCount {
+    failed: number,
+    refused: number,
+    succeeded: number
+  }
+
+  interface StatsResourceCount {
+    mean: number,
+    min: number,
+    opened: number,
+    peak: number,
+    refused: number
+  }
+
+  interface StatsConnectionTypes {
+    all: number,
+    plain: number,
+    tls: number
+  }
+
+  interface StatsMessageTraffic {
+    all: number,
+    realtime: number,
+    rest: number,
+    webhook: number
+  }
+
+  // Listeners
+  type ListenerFn = (...args: Array<any>) => void;
+
+
+  // Internal Classes
+  declare class EventEmitter {
+    on: (eventOrCallback: String | ListenerFn, callback?: ListenerFn) => void;
+
+    once: (eventOrCallback: String | ListenerFn, callback?: ListenerFn) => void;
+
+    off: (eventOrCallback?: String | ListenerFn, callback?: ListenerFn) => void;
+
+    emit: (Event: String, ...args: Array<any>) => void;
+  }
+
+  // Classes
+  export declare class Auth {
+    clientId: String;
+    authorize: (tokenParams?: TokenParams, authOptions?: AuthOptions) => TokenDetails;
+    createTokenRequest: (tokenParams?: TokenParams, authOptions?: AuthOptions) => TokenRequest;
+    requestToken: (TokenParams?: TokenParams, authOptions?: AuthOptions) => TokenDetails;
+  }
+
+  export declare class TokenDetails {
```

Comment:
Actually basically this is saying that there is a Object that follows this basic outline and has these properties.  This will allow me to take the Results of the callback and it will know that the only valid properties should be:
```
    capability: String;
    clientId: String;
    expires: number; 
    issued: number; 
    token: String;
```

Since these are pure property classes now; these actually will be converted to interfaces.  (not a lot of difference internally, but cleaner in the typings).